### PR TITLE
Add Primary key to messages

### PR DESCRIPTION
**Pull request by brunnre8**
Adds the necessary infra + runs the first bunch of migrations... namely the primary key for messages.
This will make our inserts slightly slower than now, however we can guarantee that we always have monotonically increasing IDs.

We can lower that requirement to get the speed back, with the problem that it can (and therefore will) eventually fail for someone with very fun side effects if we rely on those things in the code.

### CODE REVIEW

File: server/plugins/messageStorage/sqlite.ts
Comment (MaxLeiter) [inline server/plugins/messageStorage/sqlite.ts:31]: nit: uppercase type (`Migration`)

**Comment by brunnre8** [comment]
resolved the review comment

**Review by itsjohncs — DISMISSED** [review]

File: server/plugins/messageStorage/sqlite.ts
Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts:165]: nah, I don't think that's a good idea.
We will add incompatible changes which won't work if we add features on top.

Why do you wanna support downgrades?

Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts:165]: I'd be ok with adding "down" scripts, but I sure as hell don't add code to run with more than one schema on a version. There should be one and only one state

Comment (itsjohncs) [reply server/plugins/messageStorage/sqlite.ts:165]: > We will add incompatible changes which won't work if we add features on top.

I don't think I understand the scenario you're worried about. Could you spell out a series of events to help me?

> Why do you wanna support downgrades?

Rolling back an upgrade is what I'm concerned about. A user upgrades, something's broken that they rely on, they rollback the upgrade to wait for the next release. With the functionality I highlighted, this user would still have a broken release after rolling back.

Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts:165]: I'd rather roll forward than backwards... 
If a user really wants to downgrade they can, they simply can start over with a new db.

It's simple really, anything but the most trivial migration *will* break old code, as the semantics differ.
Say I split networks / message type to their own table, as we are storing tons of strings for no reason where an int would do.
Old code has no way of interpreting that.

Even if you have "compatible" changes, sqlite by default doesn't enforce much, meaning old code is bound to enter state which new code would disallow / fixup.
Now you constantly have to write defensive code as you may have any corrupt state based on whatever the old table default is.

In any case, if you are dead set to support downgrades, fine. I'll defer to you, but then let's do downgrade migrations.
Is that a reasonable compromise?

Comment (itsjohncs) [reply server/plugins/messageStorage/sqlite.ts:165]: > Is that a reasonable compromise?

It's reasonable but it would also be very complex to execute. So I don't think it's what we should do.

Downgrade migrations aren't straightforward. The downgrade migration code would need to be in the previous version that you're rolling back to if we followed the strategy we're using for upgrade migrations. We'd probably want to use a different strategy for migrations instead if we wanted downgrade migrations.

I'm not really against doing them but since we don't currently have any migrations that need downgrading it seems like unnecessary complexity.

> It's simple really, anything but the most trivial migration will break old code, as the semantics differ.

I understand your concern now, thanks for spelling it out. This is what I meant to address by saying you should add a comment to the migrations list saying that our migration code is very simple and doesn't handle the case of migrations that aren't backwards compatible with previous versions.

This comment will (most likely) prevent us from breaking anything in a future migration accidentally.

--

Here's my argument for removing your throw:

* The throw is currently dead code. It won't do anything to prevent downgrades from the migration you added because the throw won't exist on the version the user is downgrading to.
* The only migration currently is backwards compatible.
* Quite complex migrations can be made to be backwards incompatible: it's not that lofty a goal.
* A comment in the migration list should be sufficient to prevent accidental breakages.

But I don't contribute much to this project currently so I don't actually have much stake and I'll defer to ya'll. I'll remove my blocking review.

Comment (itsjohncs) [reply server/plugins/messageStorage/sqlite.ts:165]: > I'll remove my blocking review.

Actually I have no idea how to do remove a request for changes without explicitly approving so I'll leave the state as-is.

Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts:165]: >The throw is currently dead code. It won't do anything to prevent downgrades from the migration you added because the throw won't exist on the version the user is downgrading to.
The only migration currently is backwards compatible.

Yes, that's by design 😉

The point is that we now add infrastructure that can in the future do stuff.
That's the reason I'm not actually doing anything with the new field.


>Quite complex migrations can be made to be backwards incompatible: it's not that lofty a goal.

Frankly, if you want to do anything at all you need to break backwards compatibility.
That's why you run migrations.


>A comment in the migration list should be sufficient to prevent accidental breakages.

But i want to break it, adding features on top that can't be done as is with the current schema is the point.

>It's reasonable but it would also be very complex to execute. So I don't think it's what we should do.

It's not really.
I can add a downmigration table that stores down migrations inside of the DB.
As the only thing we in the next release do is not used that's alright if a user downgrades now.
Anything in the future would be handled correctly unless they migrate to the version we have now.
But frankly why should they?
At most you'd downgrade one version if we really messed up. Which should not happen in the first place.

Comment (itsjohncs) [reply server/plugins/messageStorage/sqlite.ts:165]: > Frankly, if you want to do anything at all you need to break backwards compatibility.
> That's why you run migrations.

I don't want to push any further on the argument I've been making. But I do want to at least impart some cool knowledge: you really can make almost any migration backwards compatible. It can take some consideration, but in most cases it's really not bad.

Basically every migration in any even-somewhat-major website is backwards compatible, since you need to be able to rollback when the alternative is having an outage for awhile. I'm sure there's some cool blog posts about how to do it in various cases if you're interested. I bet ChatGPT could talk about it too 😂. That thing is pretty good at exploratory stuff like this.

Comment (brunnre8) [reply server/plugins/messageStorage/sqlite.ts:165]: Then... how do you extract stuff to their own tables?
You can't in a compatible way.

And you want to, depending on your trade offs.

We know have a giant large table, it doesn't scale.
Even a 7-8GB DB crawls to a halt on certain operations say search (as we are doing a linear search with a very inefficient query)

**Comment by brunnre8** [comment]
IRC discussion

23:27 <+johncs> How would you implement a down migration for the lounge?
23:27 <+johncs> (my question is an attempt to answer your question, since my previous attempt did not work)
23:28 <+bookworm> easy enough, store it in the db. Play back everything > current version 
23:28 <+bookworm> there's a single place of failure, version == now()
23:28 <+johncs> Could you be more precise about what you mean? I don't understand your implementation yet
23:29 <+johncs> And specifically where is the code that knows how to do the down migration?
23:29 <+johncs> (where would the code be I mean, I know it doesn't currently exist)
23:29 <+bookworm> you add 2 tables, migration state + migration steps (so that you have the nested array our migrations in the upward directions are now). You make a query to that table with the current version. Anything higher you roll back by executing all migration steps in order on the DB
23:29 *** kimbro (~kimbro@2600:1700:3bd3:4830:213:ceff:fe0b:79e4) (kimbro) has joined the channel
23:30 <+bookworm> you drip tables, recreate tables as necessary 
23:30 <+johncs> Ah I see, you're storing the down migration steps in the DB? Like as a list of instructions? "Run all the queries in this table in order to downgrade"
23:30 <+bookworm> yes
23:31 <+bookworm> it fails exactly when you rollback to a version without the downmigration logic, which would be the current now
23:31 <+johncs> Yeah that sounds great. I understand your implementation now and I think that would work fine. I have no problems.
23:31 <+bookworm> ack, will do

**Review by MaxLeiter — APPROVED** [review]
Approving because the code looks good and runs for me, but unsure on where you and @itsjohncs left your discussion. Thank you both for your efforts!

**Comment by brunnre8** [comment]
@itsjohncs would you be ok if I open a new issue as a todo for the down migrations, but merge this as is with the hard failure for now?
Or do you want to see this working first?

**Comment by itsjohncs** [comment]
I still think it's a bad idea but I'm not blocking you.

My suggestion that I don't really wanna debate, but if it sounds good you're welcome to use it: remove the hard failure, make a comment next to the schema number that nothing is preventing rollbacks from occurring so if someone wants to make a non-roll-backable schema change they ought to put in the work then to figure out what to do.

**Comment by progval** [comment]
> This will make our inserts slightly slower than now, however we can guarantee that we always have monotonically increasing IDs.

Are you sure about that? According to [the sqlite documentation](https://www.sqlite.org/rowidtable.html), this `id INTEGER PRIMARY KEY AUTOINCREMENT` column would be an alias of the already-existing internal `rowid`, so it won't change much, as far as I understand.

**Comment by brunnre8** [comment]
yeah https://www.sqlite.org/autoinc.html

Autoincrement is the problem, not the index.
But if we allow id reuse we're in a world of pain if we try to reason about ids and fetch messages around some context, which we want to do

**Pull request merged**
